File: python/lesson-08-fastapi/src/user_service.py
"""In-memory user service — dependency injected into routes."""

from models import CreateUserRequest, UpdateUserRequest, UserResponse


class UserService:
    def __init__(self) -> None:
        self._store: dict[int, dict] = {
            1: {"id": 1, "name": "Alice", "email": "alice@example.com"},
            2: {"id": 2, "name": "Bob",   "email": "bob@example.com"},
        }
        self._next_id = 3

    def get_all(self) -> list[UserResponse]:
        return [UserResponse(**u) for u in self._store.values()]

    def get_by_id(self, user_id: int) -> UserResponse | None:
        if user_id not in self._store:
            return None
        return UserResponse(**self._store[user_id])

    def create(self, req: CreateUserRequest) -> UserResponse:
        user = {"id": self._next_id, "name": req.name, "email": req.email}
        self._store[self._next_id] = user
        self._next_id += 1
        return UserResponse(**user)

    def update(self, user_id: int, req: UpdateUserRequest) -> UserResponse | None:
        if user_id not in self._store:
            return None
        existing = self._store[user_id]
        if req.name is not None:
            existing["name"] = req.name
        if req.email is not None:
            existing["email"] = req.email
        return UserResponse(**existing)

    def delete(self, user_id: int) -> bool:
        if user_id not in self._store:
            return False
        del self._store[user_id]
        return True

File: dotnet/lesson-08-http-api/src/UserService.cs
namespace Lesson08;

public interface IUserService
{
    IEnumerable<User> GetAll();
    User? GetById(int id);
    User Create(CreateUserRequest req);
    User? Update(int id, UpdateUserRequest req);
    bool Delete(int id);
}

public class InMemoryUserService : IUserService
{
    private readonly Dictionary<int, User> _store = new()
    {
        [1] = new User(1, "Alice", "alice@example.com"),
        [2] = new User(2, "Bob",   "bob@example.com"),
    };
    private int _nextId = 3;

    public IEnumerable<User> GetAll() => _store.Values;

    public User? GetById(int id) =>
        _store.TryGetValue(id, out var u) ? u : null;

    public User Create(CreateUserRequest req)
    {
        var user = new User(_nextId++, req.Name, req.Email);
        _store[user.Id] = user;
        return user;
    }

    public User? Update(int id, UpdateUserRequest req)
    {
        if (!_store.TryGetValue(id, out var existing)) return null;
        var updated = existing with
        {
            Name  = req.Name  ?? existing.Name,
            Email = req.Email ?? existing.Email,
        };
        _store[id] = updated;
        return updated;
    }

    public bool Delete(int id) => _store.Remove(id);
}
